Clean views created inside no_grad that are modified inplace

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32839 Clean views created inside no_grad that are modified inplace**

As mentioned in the updated comment in `variable.h`, this disambiguate code like:
```python
base = torch.rand(10, requires_grad=True)
with torch.no_grad():
    view = base[1]
view.copy_(var)
torch.autograd.grad(base.sum(), var)  # <- what should it return?
```
and the similar behavior that comes from custom Function using no_grad.

Check the comment in `variable.h` for the details of all the changes.
The changes wrt the current behavior that this introduces are:
- The pattern above used to raise "a leaf that requires grad has been modified inplace" will now raise a nice error about view in no_grad + inplace in grad mode.
- Add warning when custom Function return views that are then modified inplace. This leads to the backward not being called, or called with partial gradient. This is will become an error after 1.6
- Custom Function that return multiple view will now raise a nice error instead of and internal assert.
- Extend the allow_rebase scope to also indirect rebase: if the base or another view is modified, next time the `.grad_fn()` is accessed, the history is rebased. We detect this and adapt the warning/error message. This replaces an internal assert.
- Custom Function that return an input as-is without changing it inplace does `as_view()` on it to get a new Tensor on which to attach the grad_fn. This is now done in no_grad mode to mimic the forward method. This change should not be visible for the user, it is mostly needed for the detection code above to work properly in this case.

This replaces https://github.com/pytorch/pytorch/pull/26607 
cc @sublee

Differential Revision: [D19814114](https://our.internmc.facebook.com/intern/diff/D19814114)

albanD committed:
Clean views created inside no_grad that are modified inplace

[ghstack-poisoned]

albanD committed:
Update on "Clean views created inside no_grad that are modified inplace"


As mentioned in the updated comment in `variable.h`, this disambiguate code like:
```python
base = torch.rand(10, requires_grad=True)
with torch.no_grad():
    view = base[1]
view.copy_(var)
torch.autograd.grad(base.sum(), var)  # <- what should it return?
```
Given that there is no consensus of what should happen here (does the gradient flow through the view in the no_grad or not). This special case is detected and forbidden.
As mentionned in the error message:
- If you want it to be tracked: move both out of the no_grad
- If do not want them to be tracked, move both inside the no_grad


This replaces https://github.com/pytorch/pytorch/pull/26607 
cc @sublee 

[ghstack-poisoned]

albanD committed:
Update on "Clean views created inside no_grad that are modified inplace"


As mentioned in the updated comment in `variable.h`, this disambiguate code like:
```python
base = torch.rand(10, requires_grad=True)
with torch.no_grad():
    view = base[1]
view.copy_(var)
torch.autograd.grad(base.sum(), var)  # <- what should it return?
```
Given that there is no consensus of what should happen here (does the gradient flow through the view in the no_grad or not). This special case is detected and forbidden.
As mentionned in the error message:
- If you want it to be tracked: move both out of the no_grad
- If do not want them to be tracked, move both inside the no_grad


This replaces https://github.com/pytorch/pytorch/pull/26607 
cc @sublee 

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 687715dd:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=687715dd577e200c6e3b17dbf058e1165c94fc85).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4384297) pytorch_linux_backward_compatibility_check_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=93480615) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=93480615))
<details>
<summary>
<code>Jan 30 22:21:13 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not.</code>
</summary>

```
Jan 30 22:21:13 processing existing schema:  aten::values(Dict(str, t) self) -> (t[](*)) 
Jan 30 22:21:13 processing existing schema:  aten::values(Dict(int, t) self) -> (t[](*)) 
Jan 30 22:21:13 processing existing schema:  aten::values(Dict(float, t) self) -> (t[](*)) 
Jan 30 22:21:13 processing existing schema:  aten::values(Dict(Tensor, t) self) -> (t[](*)) 
Jan 30 22:21:13 processing existing schema:  aten::view(Tensor(a) self, int[] size) -> (Tensor(a)) 
Jan 30 22:21:13 processing existing schema:  aten::where(Tensor condition) -> (Tensor[]) 
Jan 30 22:21:13 processing existing schema:  aten::where.self(Tensor condition, Tensor self, Tensor other) -> (Tensor) 
Jan 30 22:21:13 processing existing schema:  aten::zeros_like(Tensor self, *, int? memory_format=None) -> (Tensor) 
Jan 30 22:21:13 processing existing schema:  aten::zeros_like.dtype(Tensor self, *, int dtype, int layout, Device device, bool pin_memory=False, int? memory_format=None) -> (Tensor) 
Jan 30 22:21:13 processing existing schema:  prim::Constant() -> (...) 
Jan 30 22:21:13 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not.  
Jan 30 22:21:13  
Jan 30 22:21:13 Broken ops: [ 
Jan 30 22:21:13 	aten::join(str self, str[] values) -> (str) 
Jan 30 22:21:13 ] 
Jan 30 22:21:13 + cleanup 
Jan 30 22:21:13 + retcode=1 
Jan 30 22:21:13 + set +x 
Jan 30 22:21:13 =================== sccache compilation log =================== 
Jan 30 22:21:13 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 30 22:21:13 Compile requests                 0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

albanD committed:
Update on "Clean views created inside no_grad that are modified inplace"


As mentioned in the updated comment in `variable.h`, this disambiguate code like:
```python
base = torch.rand(10, requires_grad=True)
with torch.no_grad():
    view = base[1]
view.copy_(var)
torch.autograd.grad(base.sum(), var)  # <- what should it return?
```
Given that there is no consensus of what should happen here (does the gradient flow through the view in the no_grad or not). This special case is detected and forbidden.
As mentionned in the error message:
- If you want it to be tracked: move both out of the no_grad
- If do not want them to be tracked, move both inside the no_grad


This replaces https://github.com/pytorch/pytorch/pull/26607 
cc @sublee 

[ghstack-poisoned]

albanD committed:
Update on "Clean views created inside no_grad that are modified inplace"


As mentioned in the updated comment in `variable.h`, this disambiguate code like:
```python
base = torch.rand(10, requires_grad=True)
with torch.no_grad():
    view = base[1]
view.copy_(var)
torch.autograd.grad(base.sum(), var)  # <- what should it return?
```
Given that there is no consensus of what should happen here (does the gradient flow through the view in the no_grad or not). This special case is detected and forbidden.
As mentionned in the error message:
- If you want it to be tracked: move both out of the no_grad
- If do not want them to be tracked, move both inside the no_grad

This implies that any custom Function that returns views does not allow inplace modification on its output. I'll add a PR to the stack to relax this to be a DeprecationWarning for now. And we will make it into an actual error for 1.6

This replaces https://github.com/pytorch/pytorch/pull/26607 
cc @sublee 

[ghstack-poisoned]

albanD committed:
Update on "Clean views created inside no_grad that are modified inplace"


As mentioned in the updated comment in `variable.h`, this disambiguate code like:
```python
base = torch.rand(10, requires_grad=True)
with torch.no_grad():
    view = base[1]
view.copy_(var)
torch.autograd.grad(base.sum(), var)  # <- what should it return?
```
Given that there is no consensus of what should happen here (does the gradient flow through the view in the no_grad or not). This special case is detected and forbidden.
As mentionned in the error message:
- If you want it to be tracked: move both out of the no_grad
- If do not want them to be tracked, move both inside the no_grad

This implies that any custom Function that returns views does not allow inplace modification on its output. I'll add a PR to the stack to relax this to be a DeprecationWarning for now. And we will make it into an actual error for 1.6

This replaces https://github.com/pytorch/pytorch/pull/26607 
cc @sublee 

[ghstack-poisoned]

albanD committed:
Update on "Clean views created inside no_grad that are modified inplace"


As mentioned in the updated comment in `variable.h`, this disambiguate code like:
```python
base = torch.rand(10, requires_grad=True)
with torch.no_grad():
    view = base[1]
view.copy_(var)
torch.autograd.grad(base.sum(), var)  # <- what should it return?
```
Given that there is no consensus of what should happen here (does the gradient flow through the view in the no_grad or not). This special case is detected and forbidden.
As mentionned in the error message:
- If you want it to be tracked: move both out of the no_grad
- If do not want them to be tracked, move both inside the no_grad

This implies that any custom Function that returns views does not allow inplace modification on its output. I'll add a PR to the stack to relax this to be a DeprecationWarning for now. And we will make it into an actual error for 1.6

This replaces https://github.com/pytorch/pytorch/pull/26607 
cc @sublee 

[ghstack-poisoned]

gchanan commented:
> I'll add a PR to the stack to relax this to be a DeprecationWarning for now. And we will make it into an actual error for 1.6

is this referring to the example above or to the specific case of "any custom Function that returns views does not allow inplace modification on its output" ?

I don't really understand the choice of banning an operation because we can't come to a consensus about what to do with it.

albanD commented:
@gchanan 

> is this referring to the example above or to the specific case of "any custom Function that returns views does not allow inplace modification on its output" ?

Any custom function that returns views does not allow inplace modification on the result. This is because such modification would rewrite the graph and either break the graph (if the base is a temporary variable) or remove the user-defined backward (and assume the backward was only containing view operations). Since the (only) use-case for custom Function is for the user to define their own backward, this is most likely not expected by the user and should raise a nice error.

> I don't really understand the choice of banning an operation because we can't come to a consensus about what to do with it.

The problem is that different people expect different result from this code. And I cannot find a good justification for one side or the other. In particular, when asked, people either have no idea what this is doing, expect the gradients to be tracked or expect the gradients not to be tracked (which is not what we do today).
Moreover, moving both lines either inside the no grad block or both outside make the intent of the user clear.
So it is always better not to use this code pattern. So we forbid it.

gchanan commented:
> Any custom function that returns views does not allow inplace modification on the result. This is because such modification would rewrite the graph and either break the graph (if the base is a temporary variable) or remove the user-defined backward (and assume the backward was only containing view operations). Since the (only) use-case for custom Function is for the user to define their own backward, this is most likely not expected by the user and should raise a nice error.

Sorry, I don't understand the answer to my question.  My question was *what exactly* will be a DeprecationWarning in 1.5?

gchanan commented:
> The problem is that different people expect different result from this code. And I cannot find a good justification for one side or the other. In particular, when asked, people either have no idea what this is doing, expect the gradients to be tracked or expect the gradients not to be tracked (which is not what we do today).
Moreover, moving both lines either inside the no grad block or both outside make the intent of the user clear.
So it is always better not to use this code pattern. So we forbid it.

Sure, but having a warning that says this pattern is undefined behavior (you can say what we do today if you want) has the same properties.

albanD commented:
> My question was what exactly will be a DeprecationWarning in 1.5?

Modifying inplace a view created by a custom Function.

>Sure, but having a warning that says this pattern is undefined behavior (you can say what we do today if you want) has the same properties.

I don't see the point of having a warning rather than an error here.
Allowing this does not allow the user to do more stuff. Running the same code without the `torch.no_grad()` will have the exact same behavior but will be clearer.

albanD committed:
Update on "Clean views created inside no_grad that are modified inplace"


As mentioned in the updated comment in `variable.h`, this disambiguate code like:
```python
base = torch.rand(10, requires_grad=True)
with torch.no_grad():
    view = base[1]
view.copy_(var)
torch.autograd.grad(base.sum(), var)  # <- what should it return?
```
Given that there is no consensus of what should happen here (does the gradient flow through the view in the no_grad or not). This special case is detected and forbidden.
As mentionned in the error message:
- If you want it to be tracked: move both out of the no_grad
- If do not want them to be tracked, move both inside the no_grad

This implies that any custom Function that returns views does not allow inplace modification on its output. I'll add a PR to the stack to relax this to be a DeprecationWarning for now. And we will make it into an actual error for 1.6

This replaces https://github.com/pytorch/pytorch/pull/26607 
cc @sublee

Differential Revision: [D19814114](https://our.internmc.facebook.com/intern/diff/D19814114)

[ghstack-poisoned]

albanD committed:
Update on "Clean views created inside no_grad that are modified inplace"


As mentioned in the updated comment in `variable.h`, this disambiguate code like:
```python
base = torch.rand(10, requires_grad=True)
with torch.no_grad():
    view = base[1]
view.copy_(var)
torch.autograd.grad(base.sum(), var)  # <- what should it return?
```
Given that there is no consensus of what should happen here (does the gradient flow through the view in the no_grad or not). This special case is detected and forbidden.
As mentionned in the error message:
- If you want it to be tracked: move both out of the no_grad
- If do not want them to be tracked, move both inside the no_grad

This implies that any custom Function that returns views does not allow inplace modification on its output. I'll add a PR to the stack to relax this to be a DeprecationWarning for now. And we will make it into an actual error for 1.6

This replaces https://github.com/pytorch/pytorch/pull/26607 
cc @sublee

Differential Revision: [D19814114](https://our.internmc.facebook.com/intern/diff/D19814114)

[ghstack-poisoned]

ezyang commented:
> I don't see the point of having a warning rather than an error here.

Greg's point is that there is probably user code out there which is working based on the current behavior, that you will purposely break them with this diff. That's the very definition of a BC-breaking change. Now, if it had been the case that it was silently wrong and the user didn't realize they were wrong, throwing an error in those situations is a good thing, but I don't feel like that's the case here.

albanD commented:
Sounds good. I'll merge this PR with the next one and add different types of warnings.
Is it fine to deprecate the `Function` behavior (from the next PR) in 1.6 (this one causes silently wrong gradients)?
For this one (all other use cases), is 1.6 too early? We can push it back a bit more as it does not lead to always wrong, only potentially unexpected result.

ezyang commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -42,18 +42,27 @@ namespace torch { namespace autograd {
+
+      }
+    }
+    if (var.is_leaf()) {
```
Can a leaf be a view?!

albanD replied:
This might require later fix:
```python
import torch

a = torch.rand(10, requires_grad=True)

with torch.no_grad():
    b = a.narrow(0, 0, 5)
    print("b is a leaf that requires grad!: ", b.requires_grad, b.is_leaf)

c = b.sum()
print("Out of place op does create the graph")
print("c requires grad and has a grad_fn", c.requires_grad, c.grad_fn)

c.backward()
print(b.grad)
print("But the backward does not populate b.grad!", b.grad)

print("Because the accumulate grad for b does not exists:", c.grad_fn.next_functions)
```
Outputs:
```
b is a leaf that requires grad!:  True True
Out of place op does create the graph
c requires grad and has a grad_fn True <SumBackward0 object at 0x107569d50>
None
But the backward does not populate b.grad! None
Because the accumulate grad for b does not exists: ((None, 0),)
```

ezyang commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -42,18 +42,27 @@ namespace torch { namespace autograd {
-          " is being modified inplace but this is not allowed as it would prevent correct gradient computation.");
+      if (grad_fn) {
+        TORCH_CHECK(diff_view_meta->allow_rebase_history,
+            "The ", diff_view_meta->output_nr_, "th output of ", grad_fn->name(),
```
You're going to end up with error messages like "The 1th output of". Better to reword it as something like "Output 1 of"

albanD replied:
Changed it.

ezyang commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -42,18 +42,27 @@ namespace torch { namespace autograd {
+            "A view created in no_grad mode is being modified inplace but this is not allowed as the expected"
+            " behavior is unclear. You should have both the view and inplace inside the no_grad block if you"
+            " do NOT want the change to be tracked by the autograd. Or both outside the no_grad block if you"
+            " want the change to tracked.");
```
It seems to me that it won't always be obvious to a user how to move the view and inplace to the `no_grad` block. It would be a lot easier if you could tell when the view was created, but we don't track that information.

albanD replied:
In anomaly mode, we could get stack traces where the view was created.
Do you think another PR adding this would be useful?

ezyang replied:
I'd probably be lazy about this... something like, "We could give you more detailed stack traces, but we didn't implement it. Holler in the bug issue tracker if you would have liked this!"

albanD committed:
Update on "Clean views created inside no_grad that are modified inplace"


As mentioned in the updated comment in `variable.h`, this disambiguate code like:
```python
base = torch.rand(10, requires_grad=True)
with torch.no_grad():
    view = base[1]
view.copy_(var)
torch.autograd.grad(base.sum(), var)  # <- what should it return?
```
Given that there is no consensus of what should happen here (does the gradient flow through the view in the no_grad or not). This special case is detected and forbidden.
As mentionned in the error message:
- If you want it to be tracked: move both out of the no_grad
- If do not want them to be tracked, move both inside the no_grad

This implies that any custom Function that returns views does not allow inplace modification on its output. I'll add a PR to the stack to relax this to be a DeprecationWarning for now. And we will make it into an actual error for 1.6

This replaces https://github.com/pytorch/pytorch/pull/26607 
cc @sublee

Differential Revision: [D19814114](https://our.internmc.facebook.com/intern/diff/D19814114)

[ghstack-poisoned]

albanD committed:
Update on "Clean views created inside no_grad that are modified inplace"


As mentioned in the updated comment in `variable.h`, this disambiguate code like:
```python
base = torch.rand(10, requires_grad=True)
with torch.no_grad():
    view = base[1]
view.copy_(var)
torch.autograd.grad(base.sum(), var)  # <- what should it return?
```
Given that there is no consensus of what should happen here (does the gradient flow through the view in the no_grad or not). This special case is detected and forbidden.
As mentionned in the error message:
- If you want it to be tracked: move both out of the no_grad
- If do not want them to be tracked, move both inside the no_grad

This implies that any custom Function that returns views does not allow inplace modification on its output. I'll add a PR to the stack to relax this to be a DeprecationWarning for now. And we will make it into an actual error for 1.6

This replaces https://github.com/pytorch/pytorch/pull/26607 
cc @sublee

Differential Revision: [D19814114](https://our.internmc.facebook.com/intern/diff/D19814114)

[ghstack-poisoned]

albanD committed:
Update on "Clean views created inside no_grad that are modified inplace"


As mentioned in the updated comment in `variable.h`, this disambiguate code like:
```python
base = torch.rand(10, requires_grad=True)
with torch.no_grad():
    view = base[1]
view.copy_(var)
torch.autograd.grad(base.sum(), var)  # <- what should it return?
```
and the similar behavior that comes from custom Function using no_grad.

Check the comment in `variable.h` for the details of all the changes.
The changes wrt the current behavior that this introduces are:
- The pattern above used to raise "a leaf that requires grad has been modified inplace" will now raise a nice error about view in no_grad + inplace in grad mode.
- Add warning when custom Function return views that are then modified inplace. This leads to the backward not being called, or called with partial gradient. This is will become an error after 1.6
- Custom Function that return multiple view will now raise a nice error instead of and internal assert.
- Extend the allow_rebase scope to also indirect rebase: if the base or another view is modified, next time the `.grad_fn()` is accessed, the history is rebased. We detect this and adapt the warning/error message.
- Custom Function that return an input as-is without changing it inplace does `as_view()` on it to get a new Tensor on which to attach the grad_fn. This is now done in no_grad mode to mimic the forward method.

This replaces https://github.com/pytorch/pytorch/pull/26607 
cc @sublee

Differential Revision: [D19814114](https://our.internmc.facebook.com/intern/diff/D19814114)

[ghstack-poisoned]

albanD committed:
Update on "Clean views created inside no_grad that are modified inplace"


As mentioned in the updated comment in `variable.h`, this disambiguate code like:
```python
base = torch.rand(10, requires_grad=True)
with torch.no_grad():
    view = base[1]
view.copy_(var)
torch.autograd.grad(base.sum(), var)  # <- what should it return?
```
and the similar behavior that comes from custom Function using no_grad.

Check the comment in `variable.h` for the details of all the changes.
The changes wrt the current behavior that this introduces are:
- The pattern above used to raise "a leaf that requires grad has been modified inplace" will now raise a nice error about view in no_grad + inplace in grad mode.
- Add warning when custom Function return views that are then modified inplace. This leads to the backward not being called, or called with partial gradient. This is will become an error after 1.6
- Custom Function that return multiple view will now raise a nice error instead of and internal assert.
- Extend the allow_rebase scope to also indirect rebase: if the base or another view is modified, next time the `.grad_fn()` is accessed, the history is rebased. We detect this and adapt the warning/error message. This replaces an internal assert.
- Custom Function that return an input as-is without changing it inplace does `as_view()` on it to get a new Tensor on which to attach the grad_fn. This is now done in no_grad mode to mimic the forward method. This change should not be visible for the user, it is mostly needed for the detection code above to work properly in this case.

This replaces https://github.com/pytorch/pytorch/pull/26607 
cc @sublee

Differential Revision: [D19814114](https://our.internmc.facebook.com/intern/diff/D19814114)

[ghstack-poisoned]

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -782,9 +782,13 @@ def wrap_output(call):
+                # If we are in a no grad block, raise a warning
+                allow_rebase = "OnRebase::ALLOW_REBASE"
+                warn_rebase = "OnRebase::WARN_REBASE"
+                error_rebase = "OnRebase::ERROR_REBASE"
```
nit: Not sure these variables are actually buying you all that much

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -782,9 +782,13 @@ def wrap_output(call):
+                allow_rebase = "OnRebase::ALLOW_REBASE"
+                warn_rebase = "OnRebase::WARN_REBASE"
+                error_rebase = "OnRebase::ERROR_REBASE"
+                allow_rebase_history = "GradMode::is_enabled() ? {}: {}".format(allow_rebase, warn_rebase)
```
nit: Probably good to have a refer to the relevant Note here

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -109,10 +114,38 @@ variable_list _wrap_outputs(const variable_list &input_vars,
+    // For deprecation cycle. Can be removed after 1.6. In the case where we detected a view
+    // in no grad mode during the forward, only warn the user (do not change the flag if we
+    // return and input that is a view as is).
+    // See NOTE [ View + Inplace detection ] for why we replace everything by a warning.
```
Do you know if any of the torchvision ops using this are affected by this? Shouldn't be too many users here.

albanD replied:
None one them use `.mark_dirty()` so, `is_modified` is always false here.
And after checking them, they don't ever make views (extract data_ptr and write to the result directly).

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -324,6 +338,45 @@ struct TORCH_API AutogradMeta : public c10::AutogradMetaInterface {
+/// We want to detect views followed by inplace as they are often forbidden to ensure
+/// correctness of the computed gradients. But since we want to only notify the user
+/// when both happen, we tag the DifferentiableViewMeta when the view is created
+/// via the `make_vairable_*_view()` functions. This tag is then checked by the
```
nit: make_variable_view

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -324,6 +338,45 @@ struct TORCH_API AutogradMeta : public c10::AutogradMetaInterface {
+/// +----------------------+------------------------------------------------+------------------------------------------------+
+/// | ALLOW_REBASE         | view created in no_grad mode from a base that  | All other differentiable views                 |
+/// |                      | do not require gradients                       |                                                |
+/// +----------------------+------------------------------------------------+------------------------------------------------+
```
I'm not... sure what I'm supposed to do with this table. When would I be in a situation where I have a view meta with `WARN_REBASE` and I want to know how it was created? That seems like an unlikely question for me to ask. I'm probably more interested in the semantic reasoning behind all of this. Basically, suppose a disk crashed wiped out all of the source code, and I had to reconstruct this table from scratch. How would I reason my way there?

albanD replied:
This is the justification for the logic in `handle_view_on_rebase` in an easy to read form.
I added it here to make sure people will update it if they change this part of the code.

The semantic reasoning that lead to this is captured just above in the `Differentiable Views` note and in particular what happens when we return multiple views or what happens when the views are computed in no grad mode.

ezyang replied:
Yeah, see my comment below. It is incredibly unlikely this is going to get updated.

ezyang commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -368,4 +373,56 @@ unsigned VariableHooks::_register_hook(const Tensor& self, std::function<Tensor(
   return idx;
 }
 
+TORCH_API void handle_view_on_rebase(DifferentiableViewMeta* diff_view_meta, bool indirect) {
```
what does this function do? What is `indirect`?

albanD replied:
This is described in the header: "indirect=true means that the caller is not doing the inplace, but the inplace happened somewhere else."
Should I move the description in the cpp file?

ezyang replied:
Don't declare `TORCH_API` in the `cpp` file, it's redundant with the definition from the header.

Docs in the header is OK but this seems like a very implementation-detail specific function that I wouldn't expect as part of the public API.

albanD replied:
We need this to be available in `VariableTypeUtils`. That is why I added the `TORCH_API`. But maybe this is not needed? I will try and remove if not needed.

ezyang commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -368,4 +373,56 @@ unsigned VariableHooks::_register_hook(const Tensor& self, std::function<Tensor(
+    }
+
+    // We warn only once per view
+    diff_view_meta->allow_rebase_history = OnRebase::ALLOW_REBASE;
```
Err... this isn't thread safe, right?

ezyang replied:
(Or, this function is only ever called from inplace operations?)

albanD replied:
This is only ever called for inplace operations.
Also I did not added strict check here as the race would only make it warn twice if you modify the same Tensor inplace from two threads at the same time (sounds quite dangerous already) but can't lead to anything else.
I'll update the comment.

ezyang commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -368,4 +373,56 @@ unsigned VariableHooks::_register_hook(const Tensor& self, std::function<Tensor(
+      msg = c10::str("Output ", diff_view_meta->output_nr_, " of ", grad_fn->name(), " is a view and ",
+                     modified_obj, " modified inplace.");
+    } else {
+      msg = c10::str("A view was created in no_grad mode and ", modified_obj, " modified inplace with grad mode enable.");
```
nit: enabled

albanD committed:
Update on "Clean views created inside no_grad that are modified inplace"


As mentioned in the updated comment in `variable.h`, this disambiguate code like:
```python
base = torch.rand(10, requires_grad=True)
with torch.no_grad():
    view = base[1]
view.copy_(var)
torch.autograd.grad(base.sum(), var)  # <- what should it return?
```
and the similar behavior that comes from custom Function using no_grad.

Check the comment in `variable.h` for the details of all the changes.
The changes wrt the current behavior that this introduces are:
- The pattern above used to raise "a leaf that requires grad has been modified inplace" will now raise a nice error about view in no_grad + inplace in grad mode.
- Add warning when custom Function return views that are then modified inplace. This leads to the backward not being called, or called with partial gradient. This is will become an error after 1.6
- Custom Function that return multiple view will now raise a nice error instead of and internal assert.
- Extend the allow_rebase scope to also indirect rebase: if the base or another view is modified, next time the `.grad_fn()` is accessed, the history is rebased. We detect this and adapt the warning/error message. This replaces an internal assert.
- Custom Function that return an input as-is without changing it inplace does `as_view()` on it to get a new Tensor on which to attach the grad_fn. This is now done in no_grad mode to mimic the forward method. This change should not be visible for the user, it is mostly needed for the detection code above to work properly in this case.

This replaces https://github.com/pytorch/pytorch/pull/26607 
cc @sublee

Differential Revision: [D19814114](https://our.internmc.facebook.com/intern/diff/D19814114)

[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -368,4 +373,58 @@ unsigned VariableHooks::_register_hook(const Tensor& self, std::function<Tensor(
+
+    if (diff_view_meta->allow_rebase_history == OnRebase::ERROR_REBASE) {
+      TORCH_CHECK(false, msg, " This view is the output of a function that returns multiple views. Such functions do not"
+                         " allow the output views to be modified inplace.");
```
What should the user do instead now?

ezyang commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -368,4 +373,58 @@ unsigned VariableHooks::_register_hook(const Tensor& self, std::function<Tensor(
+    } else {
+      if (diff_view_meta->allow_rebase_history == OnRebase::WARN_REBASE) {
+        TORCH_INTERNAL_ASSERT(!grad_fn);
+        msg = c10::str(msg, " Given that this use case is ambiguous and error-prone, it is deprecated and will be forbidden"
```
I'm a user. I got this message. I am wondering, "Why is it ambiguous and error-prone"? Not explained. This might be a good thing to write in the wiki and then just link to from here.

ezyang commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -368,4 +373,58 @@ unsigned VariableHooks::_register_hook(const Tensor& self, std::function<Tensor(
+        TORCH_INTERNAL_ASSERT(!grad_fn);
+        msg = c10::str(msg, " Given that this use case is ambiguous and error-prone, it is deprecated and will be forbidden"
+                       "  starting 1.6. You can clarify your code and remove this warning by moving both the view and the"
+                       " inplace either both inside the no_grad block (if you don't want the inplace to be tracked) or both"
```
I guess one reason you wrote the table, is so that you could convince yourself that you could assume that a `no_grad` block was involved, when you end up with a `OnRebase::WARN_REBASE` setting on the enum. But this is a horribly indirect way of representing the state, when it could be represented in a much more simple way (once again, getting to the idea of thinking about the "semantics"): just straight up **record** whether or not a view was allocated in the no grad. None of this "inferring it based on the intent of the rebase warning: oh, *if* I see a `WARN_REBASE`, that must have come from a `no_grad`." What if someone else adds another site where `WARN_REBASE` is introduced? Now you have to fix this error message.

ezyang replied:
Right now, you are storing an 'operational' piece of data, whether or not to warn on rebase for a view. Instead, store a 'semantic' piece of data, e.g., where the view came from, and then interpret the semantics when it matters (when you rebase.)

albanD replied:
That sounds good. I will change it to `CreationMeta` and save flags like `DEFAULT, IN_CUSTOM_FUNCTION, MULTI_OUTPUT_CODEGEN, NO_GRAD_MODE` to make it clearer.

ezyang replied:
Sounds good to me.

albanD committed:
Update on "Clean views created inside no_grad that are modified inplace"


As mentioned in the updated comment in `variable.h`, this disambiguate code like:
```python
base = torch.rand(10, requires_grad=True)
with torch.no_grad():
    view = base[1]
view.copy_(var)
torch.autograd.grad(base.sum(), var)  # <- what should it return?
```
and the similar behavior that comes from custom Function using no_grad.

Check the comment in `variable.h` for the details of all the changes.
The changes wrt the current behavior that this introduces are:
- The pattern above used to raise "a leaf that requires grad has been modified inplace" will now raise a nice error about view in no_grad + inplace in grad mode.
- Add warning when custom Function return views that are then modified inplace. This leads to the backward not being called, or called with partial gradient. This is will become an error after 1.6
- Custom Function that return multiple view will now raise a nice error instead of and internal assert.
- Extend the allow_rebase scope to also indirect rebase: if the base or another view is modified, next time the `.grad_fn()` is accessed, the history is rebased. We detect this and adapt the warning/error message. This replaces an internal assert.
- Custom Function that return an input as-is without changing it inplace does `as_view()` on it to get a new Tensor on which to attach the grad_fn. This is now done in no_grad mode to mimic the forward method. This change should not be visible for the user, it is mostly needed for the detection code above to work properly in this case.

This replaces https://github.com/pytorch/pytorch/pull/26607 
cc @sublee

Differential Revision: [D19814114](https://our.internmc.facebook.com/intern/diff/D19814114)

[ghstack-poisoned]

albanD committed:
Update on "Clean views created inside no_grad that are modified inplace"


As mentioned in the updated comment in `variable.h`, this disambiguate code like:
```python
base = torch.rand(10, requires_grad=True)
with torch.no_grad():
    view = base[1]
view.copy_(var)
torch.autograd.grad(base.sum(), var)  # <- what should it return?
```
and the similar behavior that comes from custom Function using no_grad.

Check the comment in `variable.h` for the details of all the changes.
The changes wrt the current behavior that this introduces are:
- The pattern above used to raise "a leaf that requires grad has been modified inplace" will now raise a nice error about view in no_grad + inplace in grad mode.
- Add warning when custom Function return views that are then modified inplace. This leads to the backward not being called, or called with partial gradient. This is will become an error after 1.6
- Custom Function that return multiple view will now raise a nice error instead of and internal assert.
- Extend the allow_rebase scope to also indirect rebase: if the base or another view is modified, next time the `.grad_fn()` is accessed, the history is rebased. We detect this and adapt the warning/error message. This replaces an internal assert.
- Custom Function that return an input as-is without changing it inplace does `as_view()` on it to get a new Tensor on which to attach the grad_fn. This is now done in no_grad mode to mimic the forward method. This change should not be visible for the user, it is mostly needed for the detection code above to work properly in this case.

This replaces https://github.com/pytorch/pytorch/pull/26607 
cc @sublee

Differential Revision: [D19814114](https://our.internmc.facebook.com/intern/diff/D19814114)

[ghstack-poisoned]

albanD committed:
Update on "Clean views created inside no_grad that are modified inplace"


As mentioned in the updated comment in `variable.h`, this disambiguate code like:
```python
base = torch.rand(10, requires_grad=True)
with torch.no_grad():
    view = base[1]
view.copy_(var)
torch.autograd.grad(base.sum(), var)  # <- what should it return?
```
and the similar behavior that comes from custom Function using no_grad.

Check the comment in `variable.h` for the details of all the changes.
The changes wrt the current behavior that this introduces are:
- The pattern above used to raise "a leaf that requires grad has been modified inplace" will now raise a nice error about view in no_grad + inplace in grad mode.
- Add warning when custom Function return views that are then modified inplace. This leads to the backward not being called, or called with partial gradient. This is will become an error after 1.6
- Custom Function that return multiple view will now raise a nice error instead of and internal assert.
- Extend the allow_rebase scope to also indirect rebase: if the base or another view is modified, next time the `.grad_fn()` is accessed, the history is rebased. We detect this and adapt the warning/error message. This replaces an internal assert.
- Custom Function that return an input as-is without changing it inplace does `as_view()` on it to get a new Tensor on which to attach the grad_fn. This is now done in no_grad mode to mimic the forward method. This change should not be visible for the user, it is mostly needed for the detection code above to work properly in this case.

This replaces https://github.com/pytorch/pytorch/pull/26607 
cc @sublee

Differential Revision: [D19814114](https://our.internmc.facebook.com/intern/diff/D19814114)

[ghstack-poisoned]

zou3519 commented:
Discussed offline: This is a deprecation, not BC-breaking (yet!) so I'm removing the bc-breaking tag to better keep track of this
